libraries/AP_Compass/AP_Compass_BMM350.cpp

```diff
@@ -0,0 +1,557 @@
+ */
+bool AP_Compass_BMM350::mag_reset_and_wait()
+{
+    uint8_t data = 0;
+
+    // Get PMU command status 0 data
+    if (!read_bytes(BMM350_REG_PMU_CMD_STATUS_0, &data, 1)) {
+        return false;
+    }
+
+    // Check if power mode is normal
+    if (data & 0x08) {
+        // Set PMU command to suspend mode
+        if (!_dev->write_register(BMM350_REG_PMU_CMD, BMM350_PMU_CMD_SUSPEND_MODE)) {
+            return false;
+        }
+        hal.scheduler->delay(6);
+    }
+
+    // Set BR(bit reset) to PMU_CMD register
+    if (!_dev->write_register(BMM350_REG_PMU_CMD, BMM350_PMU_CMD_BR)) {
+        return false;
+    }
+    // Wait for sensor complete bit reset
+    hal.scheduler->delay(14);
+
+    // Get PMU command status 0 data and verify if PMU_CMD_STATUS_0 register has BR set
+    if (!read_bytes(BMM350_REG_PMU_CMD_STATUS_0, &data, 1) || (((data & 0xE0) >> 5) != 0x07)) {
+        return false;
+    }
+
+    // Set FGR(flux-guide reset) to PMU_CMD register
+    if (!_dev->write_register(BMM350_REG_PMU_CMD, BMM350_PMU_CMD_FGR)) {
+        return false;
+    }
+    hal.scheduler->delay(18);
+
+    // Get PMU command status 0 data and verify if PMU_CMD_STATUS_0 register has FGR set
+    if (!read_bytes(BMM350_REG_PMU_CMD_STATUS_0, &data, 1) || (((data & 0xE0) >> 5) != 0x05)) {
+        return false;
+    }
+
+    // Switch to normal mode
+    if (!set_power_mode(BMM350_POWER_MODE_NORMAL)) {
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * @brief Switch sensor power mode
+ */
+bool AP_Compass_BMM350::set_power_mode(const enum power_mode mode)
+{
+    uint8_t data = 0;
+
+    // Get PMU register data as current mode
+    if (!read_bytes(BMM350_REG_PMU_CMD, &data, 1)) {
+        return false;
+    }
+
+    if (data == BMM350_PMU_CMD_NORMAL_MODE || data == BMM350_PMU_CMD_UPD_OAE) {
+        // Set PMU command to suspend mode
+        if (!_dev->write_register(BMM350_REG_PMU_CMD, BMM350_POWER_MODE_SUSPEND)) {
+            return false;
+        }
+        // wait for sensor switch to suspend mode
+        hal.scheduler->delay(6);
+    }
+
+    // Set PMU command to target mode
+    if (!_dev->write_register(BMM350_REG_PMU_CMD, mode)) {
+        return false;
+    }
+
+    // Wait for mode change
+    if (mode == BMM350_POWER_MODE_NORMAL) {
+        hal.scheduler->delay(38);   // Switch from suspend mode to normal mode
+    } else {
+        hal.scheduler->delay(20);   // AVERAGING_4
+        // hal.scheduler->delay(28);   // AVERAGING_8
+    }
+
+    return true;
+}
+
+/**
+ * @brief Read bytes from sensor
+ */
+bool AP_Compass_BMM350::read_bytes(const uint8_t reg, uint8_t *out, const uint16_t read_len)
+{
+    uint8_t data[read_len + 2];
+
+    if (!_dev->read_registers(reg, data, read_len + 2)) {
+        return false;
+    }
+
+    memcpy(out, &data[2], read_len);
+
+    return true;
+}
+
+bool AP_Compass_BMM350::init()
+{
+    int8_t boot_retries = 5;
+    uint8_t chip_id = 0;
+    uint8_t data = 0;
```

Comment:
Move closer to where they're used, remove the initialisation as their value should only be checked after a successful read.